### ISSUE TITLE
User.find(...).invite! does not trigger validation?

### ISSUE THREAD

**Issue by Jacubeit** [scambra/devise_invitable#315]
Hi awesome Gem! Thanks for the work!

In a Rails4 project we have a `User` model with multiple `Roles` (rolify). One of them can (cancan) invite new Users. (Actually we need dummy records with a valid (unique) username until the user gets invited). So we use `skip_invitation: true`. So to speak: "works, yey!"...but... we also use `:confirmable` and `:skip_invitation` does not prevent the confirmable module doing its work.
IRB>> `User.invite!(email: "foo@bar.com", username: "J.D")`-> sends two emails!
IRB>> `User.invite!(email: "foo@bar.com", username: "J.D", skip_invitation: true)`-> sends a confirmation email!?????`
Whats the recommended way of doing it?
1) Will the user be confirmed after invitation, or do we have to handle it?

``` Ruby
User.invite!(email: "foo@bar.com", username: "J.D", skip_invitation: true) do |u|
  u.skip_confirmation!
end
```

2) Not a good solution, as this will enable login

``` Ruby
User.invite!(email: "foo@bar.com", username: "J.D", skip_invitation: true) do |u|
  u.confirm!
end
```

3) Custom solution
a) create User (skipping all Devise stuff, custom validations)
b) Invitable i.e. via button next to the dummy record
c) confirm! and invite (with invitation_at)

thank you!

jj

**Comment by rdosramos** [comment]
Scambra, tengo un problema usando devise_invitable, al hacer una invitación me arroja error de validación (:record can not be blank), pero si se crea el registro, y de hecho envía la invitación. Si vuelvo a hacer la misma invitación con el registro ya creado, no da mensaje de error.
En ambos casos al aceptar la invitación, la cuenta se crea sin inconvenientes.

¿alguna idea?

**Comment by scambra** [comment]
@rdosramos I would need more info, what validations you have in that model, if you have added some custom code on controller or changed any view.

**Comment by rdosramos** [comment]
Hi @scambra , the validations are:

  devise :invitable, :database_authenticatable, :registerable,
         :recoverable, :rememberable, :trackable, :validatable, :validate_on_invite => false
	validates_presence_of :Nombre, :Apellido, :Colegio_id
	validates :Colegio_id, numericality: {greater_than: 0}

In this ontroller Im overriding invitation update:

  before_filter :update_sanitized_params, only: :update

  # PUT /resource/invitation
  def update
    respond_to do |format|
      format.js do
        invitation_token = Devise.token_generator.digest(resource_class, :invitation_token, update_resource_params[:invitation_token])
        self.resource = resource_class.where(invitation_token: invitation_token).first
        resource.skip_password = true
        resource.update_attributes update_resource_params.except(:invitation_token)
      end
      format.html do
        super
      end
    end
  end
protected
  def update_sanitized_params
    devise_parameter_sanitizer.for(:accept_invitation) do |u|
      u.permit(:name, :password, :password_confirmation, :invitation_token, :Nombre, :Apellido, :Colegio_id)
    end
  end

And in the invitation 'new' view:

<%= form_for resource, :as => resource_name, :url => invitation_path(resource_name), :html => {:method => :post, :role => "form", :class => "form row"} do |f| %>
  <%= devise_error_messages! %>
  <% resource.class.invite_key_fields.each do |field| -%>
    <%= f.label field %><br />
    <%= f.email_field :email, autofocus: true, :class => "form-control", :placeholder => "Escribe el e-mail de tu amig@" %>
  <%= f.submit "Enviar invitación", :class =>  "btn btn-primary" %>
<% end %>

I also modified the 'edit' view, but I believe that works just fine, because there is no problem when accepting the invitation. I included the update override method because Im not sure if the override itself is making some issues.

I'd greatly apreciate your help.

thanks in advance.

**Comment by scambra** [comment]
With update_sanitize_params I think you shouldn't need to modify update method, also invitation is not accepted when format is js.

About record can not be blank error, I don't see where is coming from, you don't have any presence validation for :record. Also, devise_invitable doesn't validate model on invite unless you enable validate_on_invite (which you didn't)

Can you post invite_key from devise initializer?

**Comment by rdosramos** [comment]
You were rigth about there is no need to override update method. I removed it, leaving only update_sanitize_params, and acceptance works fine. Thanks for that.

But the 'can not be blank error' is still there. I've tried validate_on_invite true and false, but no diference: the errors are shown and the record is created anyway (awaiting for acceptance). Funny thing is that if I resend the invitation (when the record is already created but not yet accepted), everything is normal, no errors are shown.

The devise initializer is all as default (marked with #). Specifically for validate_on_invite:
  # Flag that force a record to be valid before being actually invited
  # Default: false
  # config.validate_on_invite = true

Should I initialize something? I've tried setting it to false, but still no diference.

Thank you for taking ypur time with this.

**Comment by scambra** [comment]
try to add these methods to your model to debug some code:

```ruby
def self.find_or_initialize_with_errors(*)
  record = super
  logger.debug "errors on find: " + record.errors.to_a.inspect
  record
end

def self._invite(*)
  super do |invitable|
    logger.debug "errors after validation: " + invitable.errors.to_a.inspect
  end
end
```

**Comment by rdosramos** [comment]
Hello Scambra, a added the code ypu suggested but Im not sure how to see the debugging result. I opened de development.log file and searched for what I think is what you need:

errors on find: ["Email is invalid"]
errors after validation: []

The last is inside the following:
Started POST "/alumnos/invitation" for 127.0.0.1 at 2015-03-10 12:37:49 -0430
Processing by Alumnos::InvitationsController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"4EMBVPWv2IPfcnz+M9AH94pAa5r6s6K2cL6qkJMHKZQ=", "alumno"=>{"email"=>"roberto2ramos@hotmail.com"}, "commit"=>"Enviar invitación"}
  [1m[35mAlumno Load (0.4ms)[0m  SELECT  "alumnos".* FROM "alumnos"  WHERE "alumnos"."id" = 21  ORDER BY "alumnos"."id" ASC LIMIT 1
  [1m[36mAlumno Load (0.4ms)[0m  [1mSELECT  "alumnos".* FROM "alumnos"  WHERE "alumnos"."email" = 'roberto2ramos@hotmail.com'  ORDER BY "alumnos"."id" ASC LIMIT 1[0m
errors on find: ["Email is invalid"]
errors after validation: []
  [1m[35mAlumno Load (0.4ms)[0m  SELECT  "alumnos".* FROM "alumnos"  WHERE "alumnos"."invitation_token" = '04143d0b0afa9f7690e5252ed7cc3607bd2261d6b2415a1215d5e831efa475f1'  ORDER BY "alumnos"."id" ASC LIMIT 1
  [1m[36m (0.1ms)[0m  [1mbegin transaction[0m
Binary data inserted for `string` type on column `invitation_token`
  [1m[35mSQL (0.5ms)[0m  INSERT INTO "alumnos" ("created_at", "email", "invitation_created_at", "invitation_sent_at", "invitation_token", "invited_by_id", "invited_by_type", "updated_at") VALUES (?, ?, ?, ?, ?, ?, ?, ?)  [["created_at", "2015-03-10 17:07:49.803576"], ["email", "roberto2ramos@hotmail.com"], ["invitation_created_at", "2015-03-10 17:07:49.802803"], ["invitation_sent_at", "2015-03-10 17:07:49.802803"], ["invitation_token", "04143d0b0afa9f7690e5252ed7cc3607bd2261d6b2415a1215d5e831efa475f1"], ["invited_by_id", 21], ["invited_by_type", "Alumno"], ["updated_at", "2015-03-10 17:07:49.803576"]]
  [1m[36mAlumno Exists (0.2ms)[0m  [1mSELECT  1 AS one FROM "alumnos"  WHERE ("alumnos"."email" = 'roberto2ramos@hotmail.com' AND "alumnos"."id" != 71) LIMIT 1[0m
  [1m[35m (216.7ms)[0m  commit transaction

**Comment by scambra** [comment]
debugging is ok, message is as expected, find return error on email because there is no user with that email, then error is cleared because email is valid, then invitable.invite! will be called

So error you get (:record can not be blank) must be added on invite! instance method
However that method doesn't run validations. Do you have some callback on that model?

**Comment by rdosramos** [comment]
You are a genius!

I did have another callback in the model that I forgot, it was a failed attempt to encript the user id in the URL, and i was creating an encripted ID value and then saving the model, it was here that the validations trigered. 

 after_create :create_encrypted_value, only: [:create]
  def create_encrypted_value
    self.encrypted_value = Digest::SHA1.hexdigest(id.to_s)
    self.save
  end

I removed that callback and it works fine now.

Muchas gracias @scambra 

I did note a diferent issue yesterday when registering new user, not invited (regular devise registration): it does not accept password input in the registration form, gives error :password can't be blank. But googling for the issue you helped me with I think I saw a post you made regarding this other issue. I'll look it up again and try to fix this one, following that post.

Again, thank you very much @scambra
